pkg/objstore/bos/bos.go

```diff
@@ -0,0 +1,392 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package bos
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"math/rand"
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/baidubce/bce-sdk-go/bce"
+	"github.com/baidubce/bce-sdk-go/services/bos"
+	"github.com/baidubce/bce-sdk-go/services/bos/api"
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+// partSize 128MB.
+const partSize = 1024 * 1024 * 128
+
+// Bucket implements the store.Bucket interface against bos-compatible(Baidu Object Storage) APIs.
+type Bucket struct {
+	logger log.Logger
+	client *bos.Client
+	name   string
+}
+
+// Config encapsulates the necessary config values to instantiate an bos client.
+type Config struct {
+	Bucket    string `yaml:"bucket"`
+	Endpoint  string `yaml:"endpoint"`
+	AccessKey string `yaml:"access_key"`
+	SecretKey string `yaml:"secret_key"`
+}
+
+func (conf *Config) validate() error {
+	if conf.Bucket == "" ||
+		conf.Endpoint == "" ||
+		conf.AccessKey == "" ||
+		conf.SecretKey == "" {
+		return errors.New("insufficient bos configuration information")
+	}
+
+	return nil
+}
+
+// parseConfig unmarshal a buffer into a Config with default HTTPConfig values.
+func parseConfig(conf []byte) (Config, error) {
+	config := Config{}
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return Config{}, err
+	}
+
+	return config, nil
+}
+
+// NewBucket new bos bucket.
+func NewBucket(logger log.Logger, conf []byte, component string) (*Bucket, error) {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	config, err := parseConfig(conf)
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing bos configuration")
+	}
+
+	if err := config.validate(); err != nil {
+		return nil, errors.Wrap(err, "validate bos configuration")
```

Comment:
All error messages need to have the form of `base verb + present participle $something`. So, it should be: `validating BOS configuration`. This way it looks nice when a couple errors are stringed together:
`creating bucket: validating BOS configuration: insufficient bos configuration information`

Reply:
Thanks, I have been revised, please review again.